l1-contracts/src/core/libraries/decoders/HeaderDecoder.sol

```diff
@@ -20,88 +20,99 @@ import {Hash} from "../Hash.sol";
+ *  |                                                                                  |              |   StateReference {
+ *  | 0x0080                                                                           | 0x20         |     l1ToL2MessageTree.root
+ *  | 0x00a0                                                                           | 0x04         |     l1ToL2MessageTree.nextAvailableLeafIndex
+ *  |                                                                                  |              |     PartialStateReference {
+ *  | 0x00a4                                                                           | 0x20         |       noteHashTree.root
+ *  | 0x00c4                                                                           | 0x04         |       noteHashTree.nextAvailableLeafIndex
+ *  | 0x00c8                                                                           | 0x20         |       nullifierTree.root
+ *  | 0x00e8                                                                           | 0x04         |       nullifierTree.nextAvailableLeafIndex
+ *  | 0x00ec                                                                           | 0x20         |       contractTree.root
+ *  | 0x010c                                                                           | 0x04         |       contractTree.nextAvailableLeafIndex
+ *  | 0x0110                                                                           | 0x20         |       publicDataTree.root
+ *  | 0x0130                                                                           | 0x04         |       publicDataTree.nextAvailableLeafIndex
+ *  |                                                                                  |              |     }
+ *  |                                                                                  |              |   }
+ *  | 0x0134                                                                           | 0x20         |   lastArchive.root
+ *  | 0x0154                                                                           | 0x04         |   lastArchive.nextAvailableLeafIndex
+ *  | 0x0158                                                                           | 0x20         |   bodyHash
+ *  |                                                                                  |              | }
  *  | ---                                                                              | ---          | ---
  */
 library HeaderDecoder {
-  // DECODING OFFSET CONSTANTS
-  // Where the start of trees metadata begins in the block
-  uint256 private constant START_TREES_BLOCK_HEADER_OFFSET = 0x80;
+  struct GlobalVariables {
+    uint256 chainId;
+    uint256 version;
+    uint256 blockNumber;
+    uint256 timestamp;
+  }
 
-  // The size of the block header elements
-  uint256 private constant TREES_BLOCK_HEADER_SIZE = 0xd8;
+  struct PartialStateReference {
```

Comment:
If we are full struct mode anyway. You could likely just add a `snapshot` struct, and then use that for all of those and for the archive.